### ISSUE TITLE
"Test: Run Test at Cursor" is too picky

### ISSUE THREAD

**Issue by stefanhaller** [microsoft/vscode#137984]
- VS Code Version: 1.62.3
- OS Version: macOS 11.6.1

For C++ tests using the C++ TestMate extension, "Test: Run Test at Cursor" only works if the cursor is on the first line of the test case at column 0. For python tests, it works if the cursor is anywhere on the first line of the test case, up to column 0 of the next line.

This is way too restrictive; if my cursor is on the first line of the test, I can just as well click the green icon that's right next to it. The point of the "Run Test at Cursor" is that it should work when my cursor is anywhere inside the test case (with the beginning of the test case potentially scrolled out of view).

And that's how it worked in Test Explorer UI; I'm missing that behavior.

Now, you might argue that it's the extension's responsibility to set the range of the TestItem appropriately to cover the whole test case. Yes, the C++ TestMate extension sets the range to `{line, 0, line, 0}`, and the Python extension to `{line, 0, line+1, 0}`, hence the behavior described above.

However, I'd argue that this is too much of a burden for extension developers. Most underlying test frameworks only report the starting lines of test cases; actually I'm not aware of a single one that reports where a test case ends. This means that all extensions would have to implement the same heuristic to extend a TestItem's range until either the next item in the same file, or the end of the file if there's no next one. I'd argue that this should be done by the testing framework.

Long story short: I'm proposing to treat the "no-man's land" between two TestItems (or after the last item) in a file as belonging to the item before it.

Does this make sense, @connor4312?

(Cc @matepek, @karthiknadig)

**Comment by connor4312** [comment]
Sure, we can use the closest test before the cursor if there's no exact range. I'm on vacation for a few weeks but PR's welcome.

Code pointer:

https://github.com/microsoft/vscode/blob/7dfcd74e6340a90b2681b10b33948014f8912643/src/vs/workbench/contrib/testing/browser/testExplorerActions.ts#L738-L754

**Comment by matepek** [comment]
> I'm proposing to treat the "no-man's land" between two TestItems (or after the last item) in a file as belonging to the item **before it**.

Seconded

**Comment by stefanhaller** [comment]
Here's a PR: #138001
